### PR TITLE
feat(lint/noBlankTarget): add allowed domains option

### PR DESCRIPTION
**Pull request by Jayllyz**
## Summary

Add #3609

## Test Plan

Added a test case.

### CODE REVIEW

**Review by chansuke — DISMISSED** [review]

File: crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs [hunk 1]
```diff
@@ -14,6 +14,9 @@ use biome_js_syntax::{
 };
 use biome_rowan::{AstNode, AstNodeList, BatchMutationExt, TriviaPieceKind};
```
[the remark this answers is not in the document]
Comment (Jayllyz) [reply crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs:16]: Tbh I just wrote it like that without even thinking 😂, I'll fix it thank you

File: crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs [hunk 2]
```diff
@@ -80,6 +83,17 @@ impl Rule for NoBlankTarget {
...
 
         if target_attribute.as_static_value()?.text() == "_blank" {
+            if !ctx.options().allow_domains.is_empty() {
+                let href_attribute = node.find_attribute_by_name("href")?;
+                if let Some(href_value) = href_attribute.as_static_value() {
+                    let href = href_value.text();
+                    let options = ctx.options();
+                    if is_allowed_domain(href, &options.allow_domains) {
+                        return None;
+                    }
```
[the remark this answers is not in the document]
Comment (Jayllyz) [reply crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs:93]: (real question) What's the benefit of doing this ?
I have very little experience with rust, so every review helps me a lot thank you!

File: crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs [hunk 3]
```diff
@@ -177,3 +191,17 @@ impl Rule for NoBlankTarget {
...
+pub struct AllowDomainOptions {
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub allow_domains: Vec<String>,
+}
+
+fn is_allowed_domain(href: &str, allow_domains: &[String]) -> bool {
+    allow_domains
+        .iter()
+        .any(|allowed| href.starts_with(allowed) || href.contains(allowed))
+}
```
[the remark this answers is not in the document]
Comment (Jayllyz) [reply crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs]: I've lint issue with this change, how should i fix it ?

File: crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chansuke) [reply crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs]: @Jayllyz 
Ok:)
With that code, I'm converting the `allow_domains` from `&[String]` to `Vec<&str>`:

`let allow_domains: Vec<&str> =
                        options.allow_domains.iter().map(AsRef::as_ref).collect();`
                        
my intention is to improve performance and here's the key differences:

`ownership`:  `&[String]` contains owned strings, while `Vec<&str>` contains references to string data owned elsewhere.

`memory Allocation`:  `&[String]` typically uses more memory due to each `String` owning its data. `Vec<&str>` is usually more memory-efficient.

`flexibility`:  `Vec<&str>` can reference string data from various sources (static strings, Strings, etc.), while &[String] specifically works with owned Strings.

`mutability`:  The contents of `Strings` in `&[String]` can be modified if the slice is mutable. `&str` in `Vec<&str>` are always immutable references. 

https://doc.rust-lang.org/std/option/enum.Option.html
https://doc.rust-lang.org/std/string/struct.String.html

File: crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chansuke) [reply crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs]: @Jayllyz 
Sorry, my bad. You can update the function:

```
fn is_allowed_domain(href: &str, allow_domains: &[&str]) -> bool {
    allow_domains
        .iter()
        .any(|allowed| href.starts_with(allowed) || href.contains(allowed))
}
```

File: crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jayllyz) [reply crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs]: Thank you very much for this explanation! I really need to take the time to learn rust in depth.

**Comment by Jayllyz** [comment]
Should I document this new options somewhere in this repo or it's in website one ?

File: crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs [hunk 4]
```diff
@@ -177,3 +195,17 @@ impl Rule for NoBlankTarget {
         ))
     }
 }
+
+#[derive(Clone, Debug, Default, Deserialize, Deserializable, Eq, PartialEq, Serialize)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct AllowDomainOptions {
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub allow_domains: Vec<String>,
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/a11y/no_blank_target.rs:204]: This option should be documented using Rust doc comments `///`. The documentation will be part of the configuration schema, used by the users

**Comment by ematipico** [comment]
> Should I document this new options somewhere in this repo or it's in website one ?

Yes please. We usually add a new header `## Options` that explains them.

File: crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap [hunk 5]
```diff
@@ -52,6 +52,18 @@ and the [the noopener documentation](https://html.spec.whatwg.org/multipage/link
 ```jsx
 <a href='http://external.link' target='_blank' rel="noopener" {...props}>child</a>
 ```
-```
 
+## Options
+
+Use the options to allow specific domains to use `target="_blank"` without `rel="noreferrer"`.
```
Comment (ematipico) [inline crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap:58]:
```suggestion
The option `allowDomains` allows specific domains to use `target="_blank"` without `rel="noreferrer"`. In the following configuration, it's allowed to use the domains `https://example.com` and `example.org`:
```

You usually want to provide an example to the users, and explain it.

File: crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap [hunk 6]
```diff
@@ -52,6 +52,18 @@ and the [the noopener documentation](https://html.spec.whatwg.org/multipage/link
...
+
+```json
+{
+"//": "...",
+"options": {
+"allowDomains": ["example.com", "example.org"]
+}
+}
+```
+```
```
Comment (ematipico) [inline crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap:68]:
``````suggestion

```json,ignore
{
	"//": "...",
		"options": {
		"allowDomains": ["https://example.com", "example.org"]
	}
}
```

```jsx,ignore
<>
	<a target="_blank" href="https://example.com"></a>
	<a target="_blank" href="example.org"></a>
<>
```

Biome doesn't check if the list contains valid URLs.

``````

I added some code snippets of how the valid JSX will look like using the options specified in the snippet. I also added a phrase at the end to set good expectations for the users. Does it look OK?

FYI I used `ignore` in the snippet, so the codegen won't try to evaluate the snippet.

**Review by ematipico — COMMENTED** [review]
Left some suggetions, let me know what you think

File: crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap (hunk 6 above)
Comment (ematipico) [reply crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap:68]: You can read more here: https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#document-the-rule

Comment (Jayllyz) [reply crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap:68]: Still getting assertion due to indentation with ignore, am I missing something?

image

Comment (ematipico) [reply crates/biome_cli/tests/snapshots/main_commands_explain/explain_valid_rule.snap:68]: Sorry, I can't understand where it's coming from. Can you provide more context?

**Review by ematipico — APPROVED** [review]

**Comment by ematipico** [comment]
Hey @chansuke could you review the PR please?

**Pull request merged**